dvc/repo/fetch.py

```diff
@@ -60,22 +60,39 @@ def fetch(
     try:
         if run_cache:
             self.stage_cache.pull(remote)
-        downloaded += self.cloud.pull(
-            used_objs,
-            jobs,
-            remote=remote,
-            show_checksums=show_checksums,
-        )
-    except NoRemoteError:
-        if not used_external and any(
-            obj.fs.scheme == Schemes.LOCAL for obj in used_objs
-        ):
-            raise
     except DownloadError as exc:
         failed += exc.amount
 
-    for repo_pair, files in used_external.items():
-        d, f = _fetch_external(self, repo_pair.url, repo_pair.rev, files, jobs)
+    external = set()
+    for odb, objs in used.items():
+        if odb is None:
+            # objs contains naive objects to be pulled from specified remote
+            d, f = _fetch_naive_objs(
+                self,
+                objs,
+                jobs=jobs,
+                remote=remote,
+                show_checksums=show_checksums,
+            )
+            downloaded += d
+            failed += f
+        elif isinstance(odb, GitObjectDB):
+            # objs contains staged import objects which should be saved
+            # last (after all other objects have been pulled)
+            external.update(objs)
```

Comment:
Is this only to keep previous behavior or is there some additional reason to keep fetching these last?

Reply:
This is to preserve the current fetch optimization for imported DVC objects and should be able to go away once pull/save are unified.

If we just call save() on the GitODB objects right now, we will end up streaming/downloading any imported DVC-tracked files, during the save(), but as we discussed it would be better to use fetch/pull to get all of the DVC-tracked files for now.